src/RealFourierArray.jl

```diff
@@ -0,0 +1,91 @@
+# Lazy structure for an array whose real and fourier part can be accessed by A.real and A.fourier
+# For normalization conventions, see PlaneWaveBasis.jl
+
+import Base.getproperty, Base.setproperty!, Base.fieldnames
+import Base.*, Base.+, Base./
+import Base.eltype
+
+## these functions are minimally optimized for the moment
+
+## TODO this always allocates both real and fourier parts, which might be a bit wasteful
+## TODO implement broadcasting. This seems to be non-trivial, so I just implement a few simple methods
+## TODO implement compressed representation in Fourier space of real arrays
+
+# Tr is the underlying real type
+# T is the type of the array in real space
+# TODO make this concrete by parametrizing on the type of the arrays
+mutable struct RealFourierArray{Tr <: Real, T <: Union{Tr, Complex{Tr}}}
+    basis::PlaneWaveBasis{Tr}
+    _real::AbstractArray{T, 3}
+    _fourier::AbstractArray{Complex{Tr}, 3}
+    _state::Int # 0: only _real set. 1: only _fourier set. 2: both set
+end
+# Type of the real part
+Base.eltype(A::RealFourierArray{Tr, T}) where {Tr, T} = T
+
+# Constructors
+RealFourierArray(basis, real, fourier) = RealFourierArray(basis, real, fourier, 2)
+function RealFourierArray(basis::PlaneWaveBasis{T}; iscomplex=false) where {T}
+    from_real(basis, zeros(iscomplex ? complex(T) : T, basis.fft_size...))
+end
+
+function from_real(basis, real_part::AbstractArray{T}) where {T <: Number}
+    RealFourierArray{real(T), T}(basis, real_part, similar(real_part, complex(T)), 0)
+end
+# this will necessarily be a real array
+function from_fourier(basis, fourier_part::AbstractArray{T}; isreal=false) where {T <: Complex}
+    if isreal
+        RealFourierArray{real(T), real(T)}(basis, similar(fourier_part, real(T)), fourier_part, 1)
+    else
+        RealFourierArray{real(T), T}(basis, similar(fourier_part), fourier_part, 1)
+    end
+end
+
+check_real(A::RealFourierArray{Tr, T}) where {Tr, T <: Real} = nothing
+check_real(A::RealFourierArray) = check_real(A.real)
+
+# lie blatantly to the user, as recommended by the docs
+Base.fieldnames(RealFourierArray) = (:basis, :real, :fourier)
+# TODO optimize this
+function Base.getproperty(A::RealFourierArray, x::Symbol)
+    if x == :real
+        if A._state == 1
+            r = G_to_r(A.basis, A._fourier)
+            if eltype(A) <: Real
+                A._real .= real(r)
+            else
+                A._real .= r
+            end
+            setfield!(A, :_state, 2)
+        end
+        return A._real
+    elseif x == :fourier
+        if A._state == 0
+            r_to_G!(A._fourier, A.basis, complex.(A._real))
+            setfield!(A, :_state, 2)
+        end
+        return A._fourier
+    else
+        getfield(A, x)
+    end
+end
+
+function Base.setproperty!(A::RealFourierArray, x)
+    error("RealFourierArray is intended to be read-only.
+(This can be bypassed by `setfield!` if you really want to).")
```

Comment:
Not properly indented.

Reply:
Intended: if I indent correctly, the error message shows up weird

Reply:
How about:
```
error("RealFourierArray is intended to be read-only. " *
      "(This can be bypassed by `setfield!` if you really want to).")
```